x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -1365,23 +1385,74 @@ private void unlink(final LFUCacheEntry entry) {
 
         /**
          * Cycles through the {@link LFUCacheEntry} from 0 to max frequency and
-         * tries to evict a chunk if no one is holding onto its resources anymore
+         * tries to evict a chunk if no one is holding onto its resources anymore.
+         *
+         * Also regularly polls for free regions and thus might steal one in case any become available.
          *
-         * @return the frequency of the evicted entry as integer or -1 if no entry was evicted from cache
+         * @return a now free IO region or null if none available.
          */
-        private int maybeEvict() {
+        private SharedBytes.IO maybeEvictAndTake(Runnable evictedNotification) {
             assert Thread.holdsLock(SharedBlobCacheService.this);
-            for (int currentFreq = 0; currentFreq < maxFreq; currentFreq++) {
-                for (LFUCacheEntry entry = freqs[currentFreq]; entry != null; entry = entry.next) {
-                    boolean evicted = entry.chunk.tryEvict();
-                    if (evicted && entry.chunk.io != null) {
-                        unlink(entry);
-                        keyMapping.remove(entry.chunk.regionKey, entry);
-                        return currentFreq;
+            long currentEpoch = epoch.get();
+            SharedBytes.IO freq0 = maybeEvictAndTakeForFrequency(evictedNotification, 0);
+            if (freq0 != null) {
+                return freq0;
+            }
+            // no frequency 0 entries, let us switch epoch and decay so we get some for next time.
+            tryNewEpoch(currentEpoch);
```

Comment:
I've been a bit confused by the comment and method, assuming the epoch would be incremented immediately, whereas the next epoch starts after decay will be computed again, which won't happen immediately due to the lock on `SharedBlobCacheService.this`

Maybe renaming to `scheduleDecayAndNextEpoch` or something along those lines?

Reply:
I called it `maybeScheduleDecayAndNextEpoch` and changed the task class name to DecayAndNextEpoch